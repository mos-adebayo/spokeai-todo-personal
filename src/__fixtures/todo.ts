export const actionItems = [
  {
    isDone: true,
    description: "description 1"
  },
  {
    isDone: false,
    description: "description 2"
  }
];

export const todo = {
  id: "1234567890",
  title: "Title",
  items: actionItems
};
