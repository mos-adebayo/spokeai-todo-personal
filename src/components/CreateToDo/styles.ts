import styled from "styled-components";

export const FormWrapper = styled.div`
  width: 50%;
  margin: 20px auto;
  
  .form-control:focus {
    box-shadow: none;
  }
`;
