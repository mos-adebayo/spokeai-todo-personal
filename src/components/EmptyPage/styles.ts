import styled from "styled-components";

export const Wrapper = styled.div`
  width: 100%;
  margin: 20px auto;
  text-align: center;

  @media screen and (min-width: 768px) {
    width: 80%;
  }

  @media screen and (min-width: 992px) {
    width: 60%;
  }

  @media screen and (min-width: 1200px) {
    width: 50%;
  }
`;

export const LinkText = styled.span`
  cursor: pointer;
  text-decoration: underline;
`;
