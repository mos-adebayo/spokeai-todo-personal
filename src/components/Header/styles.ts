import styled from "styled-components";
import { gradientPrimaryBackground } from "../../util/stylesheet";

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
  padding: 45px 20px;
  height: 100px;
  margin-bottom: 25px;
  font-family: "Mochiy Pop One", sans-serif;

  ${gradientPrimaryBackground};
`;
