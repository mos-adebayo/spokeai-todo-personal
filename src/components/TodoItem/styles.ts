import styled from "styled-components";

export const Wrapper = styled.div`
  background: #fff;
  border-radius: 12px;
  box-shadow: rgb(166 171 189 / 20%) 0 10px 20px;
  padding: 15px 30px;
  margin-bottom: 15px;
`;

export const Title = styled.h4`
color: #49486c;
`;

export const Text = styled.p`
color: #49486c
`;
