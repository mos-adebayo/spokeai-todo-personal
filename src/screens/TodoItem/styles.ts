import styled from "styled-components";

export const Wrapper = styled.div`
  width: 100%;
  margin: 0 auto;
  background: #fff;
  height: calc(100vh - 145px);

  @media screen and (min-width: 576px) {
    width: 95%;
  }

  @media screen and (min-width: 768px) {
    width: 70%;
  }

  @media screen and (min-width: 1200px) {
    width: 50%;
  }
`;
