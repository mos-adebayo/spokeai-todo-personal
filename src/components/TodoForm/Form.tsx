import React, { ChangeEvent, useState, useEffect } from "react";
import {
  Form,
  InputGroup,
  FormControl,
  Toast,
  ToastContainer
} from "react-bootstrap";
import { AddButton, FormWrapper, ItemsWrapper, ItemWrapper } from "./styles";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../redux/reducers/rootReducers";
import {
  createTaskRequest,
  createTaskStarted,
  updateTaskRequest
} from "../../redux/actions/taskActions";
import { CheckBox } from "../TodoItem/styles";

type Prop = {
  task?: TaskItemType;
};
const TaskForm: React.FC<Prop> = ({ task }) => {
  const dispatch = useDispatch();
  const { loading } = useSelector((state: RootState) => state.task);

  const [visibleToast, setVisibleToast] = useState(false);
  const [title, setTitle] = useState("");
  const [items, setItems] = useState<ActionItemPayloadType[]>([
    {
      isDone: false,
      description: "",
      isDirty: false
    }
  ]);

  useEffect(() => {
    if (task) {
      setTitle(task.title);
      const taskItems = task.items.map(({ description, isDone }) => ({
        isDone,
        description,
        isDirty: false
      }));

      const actionItems = [
        ...taskItems,
        {
          isDone: false,
          description: "",
          isDirty: false
        }
      ];

      setItems(actionItems);
    }
  }, [task]);

  const handleChange = (
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement | any>,
    item: ActionItemPayloadType,
    index: number
  ) => {
    const newItems = [...items];
    const value =
      e.target.name === "isDone" ? e.target.checked : e.target.value;
    newItems[index] = {
      ...items[index],
      [e.target.name]: value,
      isDirty: true
    };

    if (!item.isDirty && value) {
      newItems.push({
        isDone: false,
        description: "",
        isDirty: false
      });
    }

    setItems(newItems);
  };

  const handleTitleChange = (e: ChangeEvent<HTMLInputElement>) => {
    setTitle(e.target.value);
  };

  const handleCheckItem = (item: ActionItemPayloadType, index: number) => {
    if (!item.description) return;

    const newItems = [...items];
    const currentItem = items[index];
    newItems[index] = {
      ...currentItem,
      isDone: !currentItem.isDone,
      isDirty: true
    };

    setItems(newItems);
  };

  const resetForm = () => {
    handleStartCreation(false);
    setTitle("");
    setItems([
      {
        isDone: false,
        description: "",
        isDirty: false
      }
    ]);
  };

  const handleStartCreation = (status: boolean) => {
    dispatch(createTaskStarted(status));
  };
  const handleSubmit = () => {
    const actualItems = items
      .filter(({ description }) => description)
      .map(({ description, isDone }) => ({ isDone, description }));
    const payload = {
      id: task ? task.id : new Date().getTime().toString(),
      title: title,
      items: actualItems
    };

    if (task) {
      dispatch(updateTaskRequest(payload));
      setVisibleToast(true);
    } else {
      dispatch(createTaskRequest(payload));
      resetForm();
    }
  };

  return (
    <FormWrapper>
      <InputGroup className="mb-2">
        <Form.Control
          value={title}
          type="text"
          placeholder={"Title"}
          onChange={handleTitleChange}
        />
        <AddButton disabled={!title || loading} onClick={handleSubmit}>
          Save
        </AddButton>
      </InputGroup>

      <ItemsWrapper editMode={Boolean(task)}>
        {items.map((item, index) => (
          <ItemWrapper key={index}>
            <InputGroup className="mb-1" key={index}>
              <CheckBox
                checked={item.isDone}
                onClick={() => handleCheckItem(item, index)}
              />
              <FormControl
                value={item.description}
                name="description"
                placeholder="Type new task item"
                onChange={(e) => handleChange(e, item, index)}
              />
            </InputGroup>
          </ItemWrapper>
        ))}
      </ItemsWrapper>

      <ToastContainer className="p-3" position="top-end">
        <Toast
          show={visibleToast}
          onClose={() => setVisibleToast(false)}
          bg="light"
          autohide
          delay={3000}
        >
          <Toast.Body className="text-success">
            Task updated successfully!
          </Toast.Body>
        </Toast>
      </ToastContainer>
    </FormWrapper>
  );
};

export default TaskForm;
