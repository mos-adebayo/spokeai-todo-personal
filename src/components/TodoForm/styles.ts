import styled, { css } from "styled-components";
import { Button } from "react-bootstrap";
import { CheckBox } from "../TodoItem/styles";

export const AddButton = styled(Button)`
  &:focus {
    box-shadow: none;
  }

  background: linear-gradient(
    77.07deg,
    rgb(69, 183, 204) 0%,
    rgb(75, 207, 187) 100%
  );
  outline: none;
  border: none;
`;

export const AddButtonWrapper = styled.div`
  position: absolute;
  right: 50px;
  bottom: 50px;

  display: flex;
  justify-content: center;
  align-items: center;

  color: #ffffff;
  cursor: pointer;

  width: 100px;
  height: 100px;
  font-size: 80px;
  border-radius: 50%;
  background: linear-gradient(
    77.07deg,
    rgb(69, 183, 204) 0%,
    rgb(75, 207, 187) 100%
  );
  box-shadow: rgb(52 58 64 / 40%) 3px 1px 13px 0;

  @media screen and (max-width: 767px) {
    width: 80px;
    height: 80px;
    font-size: 60px;
  }
`;

export const FormWrapper = styled.div`
  margin: 20px 0;

  .form-control:focus {
    box-shadow: none;
  }
`;

export const ItemWrapper = styled.div`
  border-bottom: 1px solid #dfdfdf;
  margin: 0 -16px;
  padding: 10px 16px 8px;

  ${CheckBox} {
    margin-right: 5px;
  }

  &:first-of-type {
    border-top: 1px solid #dfdfdf;
  }
`;

const editItemWrapper = css`
  ${ItemWrapper} {
    margin: 0;
  }
`;

export const ItemsWrapper = styled.div<{ editMode?: boolean }>`
  margin-top: 15px;

  .form-control {
    height: 1.5rem;
    border: none;
  }

  ${({ editMode }) => editMode && editItemWrapper};
`;
