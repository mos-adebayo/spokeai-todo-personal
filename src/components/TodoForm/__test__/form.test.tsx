import React from "react";
import { Provider } from "react-redux";
import { mount, ReactWrapper } from "enzyme";
import Form from "../Form";
import configureStore from "redux-mock-store";
import { todo } from "../../../__fixtures/todo";
import * as redux from "react-redux";
import { AddButton, ItemWrapper } from "../styles";
import { FormControl } from "react-bootstrap";

const mockStore = configureStore();
const mockState = {
  task: { isCreating: true, loading: false, task: null }
};

describe("<TodoForm />", () => {
  let spyOnUseSelector;
  let spyOnUseDispatch;
  let wrapper: ReactWrapper;
  const mockDispatch = jest.fn();

  beforeEach(() => {
    spyOnUseSelector = jest.spyOn(redux, "useSelector");
    spyOnUseSelector.mockReturnValue(mockState);

    spyOnUseDispatch = jest.spyOn(redux, "useDispatch");
    spyOnUseDispatch.mockReturnValue(mockDispatch);

    wrapper = mount(
      <Provider store={mockStore()}>
        <Form task={todo} />
      </Provider>
    );
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("should match snapshot for form creation", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render all action items", () => {
    const items = wrapper.find(ItemWrapper);
    expect(items.length).toEqual(3);
  });

  it("should handle form inputs change", () => {
    const inputs = wrapper.find(FormControl);
    const addButton = wrapper.find(AddButton);
    expect(inputs.length).toEqual(4);

    inputs
      .at(0)
      .simulate("change", { target: { value: "title", name: "title" } });
    inputs
      .at(1)
      .simulate("change", {
        target: { value: "description", name: "description" }
      });

    addButton.simulate("click");
    expect(mockDispatch).toHaveBeenCalledTimes(1);
  });
});
