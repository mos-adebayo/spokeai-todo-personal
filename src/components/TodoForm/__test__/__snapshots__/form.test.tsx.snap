// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoForm /> should match snapshot for form creation 1`] = `
.c7 {
  border: 1px solid #283451;
  width: 1.5rem;
  height: 1.5rem;
  display: inline-block;
  margin-right: 10px;
  text-align: center;
}

.c6 {
  border: 1px solid #283451;
  width: 1.5rem;
  height: 1.5rem;
  display: inline-block;
  margin-right: 10px;
  text-align: center;
}

.c6:after {
  content: " ";
  display: inline-block;
  width: 0.5rem;
  height: 1rem;
  box-sizing: border-box;
  -webkit-transform: rotate(40deg);
  -ms-transform: rotate(40deg);
  transform: rotate(40deg);
  border: 0.1rem solid #283451;
  border-top: 0;
  border-left: 0;
}

.c1 {
  outline: none;
  border: none;
  background: linear-gradient( 77.07deg, rgb(69,183,204) 0%, rgb(75,207,187) 100% );
}

.c1:focus {
  box-shadow: none;
}

.c0 {
  margin: 20px 0;
}

.c0 .form-control:active,
.c0 .form-control:focus {
  box-shadow: none;
  border-color: #5f9ea0;
}

.c4 {
  border-bottom: 1px solid #dfdfdf;
  margin: 0 -16px;
  padding: 10px 16px 8px;
}

.c4 .c5 {
  margin-right: 5px;
}

.c4:first-of-type {
  border-top: 1px solid #dfdfdf;
}

.c2 {
  margin-top: 15px;
}

.c2 .form-control {
  height: 1.5rem;
  border: none;
}

.c2 .c3 {
  margin: 0;
}

<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <TaskForm
    task={
      Object {
        "id": "1234567890",
        "items": Array [
          Object {
            "description": "description 1",
            "isDone": true,
          },
          Object {
            "description": "description 2",
            "isDone": false,
          },
        ],
        "title": "Title",
      }
    }
  >
    <styled.div>
      <div
        className="c0"
      >
        <InputGroup
          className="mb-2"
        >
          <div
            className="mb-2 input-group"
          >
            <FormControl
              onChange={[Function]}
              placeholder="Title"
              type="text"
              value="Title"
            >
              <input
                className="form-control"
                onChange={[Function]}
                placeholder="Title"
                type="text"
                value="Title"
              />
            </FormControl>
            <Styled(Button)
              onClick={[Function]}
            >
              <Button
                active={false}
                className="c1"
                disabled={false}
                onClick={[Function]}
                variant="primary"
              >
                <button
                  className="c1 btn btn-primary"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Save
                </button>
              </Button>
            </Styled(Button)>
          </div>
        </InputGroup>
        <styled.div
          editMode={true}
        >
          <div
            className="c2"
          >
            <styled.div
              key="0"
            >
              <div
                className="c3 c4"
              >
                <InputGroup
                  className="mb-1"
                  key="0"
                >
                  <div
                    className="mb-1 input-group"
                  >
                    <styled.span
                      checked={true}
                      onClick={[Function]}
                    >
                      <span
                        checked={true}
                        className="c5 c6"
                        onClick={[Function]}
                      />
                    </styled.span>
                    <FormControl
                      name="description"
                      onChange={[Function]}
                      placeholder="Type new task item"
                      value="description 1"
                    >
                      <input
                        className="form-control"
                        name="description"
                        onChange={[Function]}
                        placeholder="Type new task item"
                        value="description 1"
                      />
                    </FormControl>
                  </div>
                </InputGroup>
              </div>
            </styled.div>
            <styled.div
              key="1"
            >
              <div
                className="c3 c4"
              >
                <InputGroup
                  className="mb-1"
                  key="1"
                >
                  <div
                    className="mb-1 input-group"
                  >
                    <styled.span
                      checked={false}
                      onClick={[Function]}
                    >
                      <span
                        checked={false}
                        className="c5 c7"
                        onClick={[Function]}
                      />
                    </styled.span>
                    <FormControl
                      name="description"
                      onChange={[Function]}
                      placeholder="Type new task item"
                      value="description 2"
                    >
                      <input
                        className="form-control"
                        name="description"
                        onChange={[Function]}
                        placeholder="Type new task item"
                        value="description 2"
                      />
                    </FormControl>
                  </div>
                </InputGroup>
              </div>
            </styled.div>
            <styled.div
              key="2"
            >
              <div
                className="c3 c4"
              >
                <InputGroup
                  className="mb-1"
                  key="2"
                >
                  <div
                    className="mb-1 input-group"
                  >
                    <styled.span
                      checked={false}
                      onClick={[Function]}
                    >
                      <span
                        checked={false}
                        className="c5 c7"
                        onClick={[Function]}
                      />
                    </styled.span>
                    <FormControl
                      name="description"
                      onChange={[Function]}
                      placeholder="Type new task item"
                      value=""
                    >
                      <input
                        className="form-control"
                        name="description"
                        onChange={[Function]}
                        placeholder="Type new task item"
                        value=""
                      />
                    </FormControl>
                  </div>
                </InputGroup>
              </div>
            </styled.div>
          </div>
        </styled.div>
        <ToastContainer
          className="p-3"
          position="top-end"
        >
          <div
            className="toast-container position-absolute top-0 end-0 p-3"
          >
            <Toast
              autohide={true}
              bg="light"
              delay={3000}
              onClose={[Function]}
              show={false}
            >
              <ToastFade
                in={false}
                unmountOnExit={true}
              >
                <Fade
                  appear={false}
                  in={false}
                  mountOnEnter={false}
                  timeout={300}
                  transitionClasses={
                    Object {
                      "entering": "showing",
                      "exiting": "showing show",
                    }
                  }
                  unmountOnExit={true}
                >
                  <ForwardRef
                    addEndListener={[Function]}
                    appear={false}
                    childRef={null}
                    in={false}
                    mountOnEnter={false}
                    onEnter={[Function]}
                    timeout={300}
                    unmountOnExit={true}
                  >
                    <Transition
                      addEndListener={[Function]}
                      appear={false}
                      enter={true}
                      exit={true}
                      in={false}
                      mountOnEnter={false}
                      nodeRef={
                        Object {
                          "current": null,
                        }
                      }
                      onEnter={[Function]}
                      onEntered={[Function]}
                      onEntering={[Function]}
                      onExit={[Function]}
                      onExited={[Function]}
                      onExiting={[Function]}
                      timeout={300}
                      unmountOnExit={true}
                    />
                  </ForwardRef>
                </Fade>
              </ToastFade>
            </Toast>
          </div>
        </ToastContainer>
      </div>
    </styled.div>
  </TaskForm>
</Provider>
`;
